core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/ExpressionEvaluator.scala

```diff
@@ -30,20 +33,258 @@ object ExpressionEvaluator {
+            - through a series of transformations, generate a Map of entity name to Map of lookup expressions and their
+              evaluated value for that entity
+            - for each entity, substitute the evaluated values of attribute references back into the input expression
+              by visiting the parsed tree of input expression
+            - for each entity, pass the reconstructed input expression to JSONEvaluator to parse the expression into
+              AttributeValue
+    To help understand the approach if their are attribute references present, we will follow the below example roughly:
+      expression = "{"exampleRef1":this.bam, "exampleIndex":this.index}"
+      rootEntities = Seq(101, 102) (here we assume the entity name is 101 for Entity Record 1 and 102 for Entity record 2)
+
+    The final output will be:
+    ReadWriteAction(Map(
+          "101" -> Try(Seq(AttributeValueRawJson("{"exampleRef1":"gs://abc", "exampleIndex":123}"))),
+          "102" -> Try(Seq(AttributeValueRawJson("{"exampleRef1":"gs://def", "exampleIndex":456}")))
+        )
+    )
+   */
+  def evalFinalAttribute(workspaceContext: SlickWorkspaceContext, expression: String)
+                        (implicit executionContext: ExecutionContext) : ReadWriteAction[Map[EntityName, Try[Iterable[AttributeValue]]]] = {
     import slickEvaluator.parser.driver.api._
 
-    JsonExpressionEvaluator.evaluate(expression) match {
-      //if the expression evals as JSON, it evaluates to the same thing for every entity, so build that map here
-      case Success(parsed) => DBIO.successful(rootEntities match {
+    /*
+      Evaluate each attribute reference expression using the Slick Evaluator.
+      @return: Sequence of tuples of attribute reference expressions and their slick evaluated value wrapped in a single DBIO action
+      For our example:
+        input: Set(this.bam, this.index)
+        output: ReadWriteAction(Seq(
+          ("this.bam", Map("101" -> Try(Seq(AttributeString("gs://abc"))), "102" -> Try(Seq(AttributeString("gs://def"))))),
+          ("this.index", Map("101" -> Try(Seq(AttributeNumber(123))), "102" -> Try(Seq(AttributeNumber(456)))))
+        ))
+     */
+    def evaluateAttributeReferences(attrRefs: Set[LookupExpression]): ReadWriteAction[Seq[(LookupExpression, Map[EntityName, Try[Iterable[AttributeValue]]])]] = {
+      val evaluatedLookupSeq: Set[ReadWriteAction[(LookupExpression, Map[EntityName, Try[Iterable[AttributeValue]]])]] = attrRefs.map { attrRef =>
+        slickEvaluator.evalFinalAttribute(workspaceContext, attrRef).map((attrRef, _))
+      }
+
+      // sequence Set[ReadWriteAction[T]] into single ReadWriteAction[Seq[T]]
+      DBIO.sequence(evaluatedLookupSeq.toSeq)
+    }
+
+    // Converts AttributeValue to JsValue (JsValue instead of String so as to preserve the evaluated value structure)
+    def unpackSlickEvaluatedOutput(value: AttributeValue): JsValue = {
```

Comment:
My understanding was that if any function (like a helper function) is going to be only used by a certain function, then it can be nested inside the top level function. I can see that `unpackSlickEvaluatedOutput` could be used elsewhere, but would that apply for other functions nested inside `evalFinalAttribute`? What is the rule of thumb here?

Reply:
I don't have a good rule. I'm leaving it to your style preference.

FYI my preference is usually to NOT nest helper functions, instead marking them `private def`, or maybe even `private[pkg] def` if one wants to unit tested the functionality.